ydb/core/tx/datashard/datashard__read_iterator.cpp

```diff
@@ -870,11 +895,15 @@ class TReader {
 
             if (ShouldStop()) {
                 LastProcessedKey = TSerializedCellVec::Serialize(rowKey.Cells());
-                LastProcessedKeyErased = false;
+                LastProcessedKeyErasedOrMissing = false;
                 return EReadStatus::StoppedByLimit;
             }
         }
 
+        if (!lastKey) {
+            lastKey = TSerializedCellVec::Serialize(iter->GetKey().Cells());
```

Comment:
Вот это возможно очень плохо. Я специально ниже брал `iter->GetKey().Cells()` чтобы не делать сериализацию без необходимости, например если мы встретили page fault без достаточного кол-ва полезной работы и будем рестартить чтение с той же позиции. У тебя же последний ключ теперь всегда сериализуется и копируется.

Reply:
Может вот так тогда?
```
  // Note: when stopping due to page faults after an erased row we will
  // reposition on that same row so erase cache can extend that cached
  // erased range. When we don't observe any user-visible rows before a
  // page fault we want to make sure we observe multiple deleted rows,
  // which must be at least 2 (because we may resume from a known deleted
  // row). When there are not enough rows we would prefer restarting in
  // the same transaction, instead of starting a new one, in which case
  // we will not update stats and will not update RowsProcessed.
  if (!lastKey) {
      auto lastKeyCells = iter->GetKey().Cells();
      if (lastKeyCells && (advanced || iter->Stats.DeletedRowSkips >= 4) && iter->Last() == NTable::EReady::Page) {
          lastKey = TSerializedCellVec::Serialize(lastKeyCells);
      }
  }

  if (lastKey) {
      LastProcessedKey = lastKey;
      LastProcessedKeyErasedOrMissing = hasMissingExternalBlobs || iter->GetKeyState() == NTable::ERowOp::Erase;
      advanced = true;
  } else {
      LastProcessedKey.clear();
  }
```

Reply:
А будет ли это всё правильно работать и так как ты ожидаешь?

Представим, что мы попадаем сразу на несколько строк (до этого не было ничего удалено и т.д.), у которых не хватает внешних блобов. Ты сделаешь проход по нескольким строкам, потрогаешь их, но в итоге у тебя окажется `lastKey = firstObservedKey`, ты выставишь `advanced = true`, итератор решит, что мы что-то полезное пропустили, но ответ пустой, выбросит текущую транзакцию (!) и пойдёт в новую транзакцию с Continue. Однако в новой транзакции всё повторится, мы ещё даже не пытались ничего грузить с диска, данных в кеше так и не появилось, у нас на первой строке не хватает блобы, мы снова выставляем lastKey и т.д.

Reply:
А ещё мне кажется тут инварианты насчёт RowsProcessed начинают нарушаться. Если был какой-то диапазон удалений, но потом всё споткнулось о внешний блоб, эта пачка удалений в RowsProcessed не попадёт (ты перестаёшь что-либо добавлять в RowsProcessed из-за continue). Если в итоге окажется, что RowsProcessed на выходе из метода нулевой, транзакция всё-таки пойдёт на рестарт, но она теперь будет позиционироваться на первую строку с внешним блобом, а на удаления смотреть не будет. Однако среди этих удалений мы могли через TxObserver видеть какие-то незакомиченные изменения, которые влияют на консистентность и конфликты. Но на повторе мы их уже не увидим, что приведёт к потенциально неконсистентному чтению.

Reply:
Впрочем LastProcessedKey переносится только если у нас есть Continue, что-то очень сложно понять будет ли оно всё правильно работать. :-/

Reply:
Может тогда можно как-то вынести код "доталкивания" итератора в конец метода IterateRange?

Тогда с виду всё останется как было, только вот надо будет добавить чтобы LastKey не обновлялся внутри итератора когда у нас "типо" пейдж-фолт на блобе

Reply:
Или вообще сделать всё как раньше (не выставлять итератору EnableReadMissingReferences), и в конце если наш пейдж-фолт на внешнем блобе "дотолкать" его уже с выставленным EnableReadMissingReferences (или даже ещё один итератор создать и его от LastKey прогнать

Reply:
Есть тест (в новых, называется ExtBlobs), я проверил, транзакция не рестартует, если в первой же строке внешний блоб, которого нет в кеше.

Насчёт второго: а нельзя RowsProcessed увеличивать в первый раз как мы встретили missing reference?

Reply:
Удаления могут быть также между строк с внешними блобами, и после них. Сейчас те что после добавятся в RowsProcessed по условию `advanced == true`, и тогда мне кажется может и в бесконечный цикл уйти (каждый раз RowsProcessed > 0, начинаем с первой строки, с диска ничего не грузим).

Reply:
Подправил + написал тесты (проверяю, что нет TxContinue, если были до/после внешних блобов удаленные строки)